### PR TITLE
Refactor cmontecarlo tests for exercising parametrization.

### PR DESCRIPTION
**Pull request by karandesai-96**
Current design procedure of tests:
- CMontecarlo methods are compiled into **montecarlo.so** (shared object)
- Corresponding tests are compiled into **test_montecarlo.so**.
- For testing these using pytest framework, **test_montecarlo.so** is wrapped and C tests are exposed to **test_cmontecarlo.py** using Python `ctypes` library.

This PR refactors the current tests such that:
- They directly wrap **montecarlo.so** and expose the C methods and struct parameters to Python instead of C tests.
- Pytest's parametrization can be used to test the methods extensively.
- Parameters of each Python test method are the ones which are used by the corresponding C method.
- RPacket and StorageModel structs are defined inside the test every time, using those parameters.

### CODE REVIEW

**Comment by karandesai-96** [comment]
Related Issues: #249 #408

**Comment by karandesai-96** [comment]
#### Parametrization of `compute_distance2boundary` is successful.

* The C method uses `r`, `mu`, `recently_crossed_boundary`, `r_inner` and `r_outer` parameters of `RPacket`/ `StorageModel`. These are passed while parametrization.
* Here is a very small snippet depicting a general design procedure:
```python
@pytest.mark.parametrize(
['mu', 'r', 'recently_crossed_boundary', 'r_inner', 'r_outer', 'expected'],
[(0.3, 7.5e14, [6.912e14, 8.64e14], [8.64e14, 1.0368e15], 259376919351035.88)])
# more test cases within
def test_compute_distance2boundary(mu, r, recently_crossed_boundary, r_inner, r_outer, expected):
    packet = RPacket(mu=mu, r=r, current_shell_id=0, next_shell_id=1,
                     recently_crossed_boundary=recently_crossed_boundary)
    model = StorageModel(r_inner=npclib.as_ctypes(np.array(r_inner)),
                         r_outer=npclib.as_ctypes(np.array(r_outer)))
    # Further assertions
```
The tests pass and everything works fine now.

File: tardis/montecarlo/struct.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from ctypes import *
+
+import numpy as np
+from tardis import __path__ as path
+
+test_path = os.path.join(path[0], 'montecarlo', 'test_montecarlo.so')
+tests = CDLL(test_path)
+
+cmontecarlo_filepath = os.path.join(path[0], 'montecarlo', 'montecarlo.so')
+cmontecarlo_methods = CDLL(cmontecarlo_filepath)
```
Comment (yeganer) [inline tardis/montecarlo/struct.py]: The loading of the CDLL is something specific to the tests, this shouldn't be in struct.py, only the definition of `RPacket` and `StorageModel`

File: tardis/montecarlo/struct.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        ('next_shell_id', c_int64),
+        ('status', c_int),
+        ('id', c_int64),
+        ('chi_th', c_double),
+        ('chi_cont', c_double),
+        ('chi_ff', c_double),
+        ('chi_bf', c_double)
+    ]
+
+    def __init__(self, **kwargs):
```
Comment (yeganer) [inline tardis/montecarlo/struct.py]: The setting of default variables is something that should be in the tests.

File: tardis/montecarlo/struct.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        ('virt_packet_last_line_interaction_in_id', POINTER(c_int64)),
+        ('virt_packet_last_line_interaction_out_id', POINTER(c_int64)),
+        ('virt_packet_count', c_int64),
+        ('virt_array_size', c_int64)
+    ]
+
+    def __init__(self, **kwargs):
+        super(StorageModel, self).__init__()
+
+        self.last_line_interaction_in_id = np.ctypeslib.as_ctypes(np.array([0] * 2))
```
Comment (yeganer) [inline tardis/montecarlo/struct.py]: Same as `RPacket`: Move the default variables to the tests.

File: tardis/montecarlo/struct.py (hunk 2 above)
Comment (karandesai-96) [inline tardis/montecarlo/struct.py]: `__init__` is inherited from `ctypes.Structure` so I had to write it within the struct declaration.

File: tardis/montecarlo/struct.py (hunk 1 above)
Comment (karandesai-96) [inline tardis/montecarlo/struct.py]: Okay, I will move it to **test_cmontecarlo.py**

File: tardis/montecarlo/struct.py (hunk 2 above)
Comment (yeganer) [inline tardis/montecarlo/struct.py]: No, you don't. You can subclass the struct and write your own init and call `super(Class, self).__init__(...)`.
The other method would be to define fixtures in the tests that return a struct initialized to the default values.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 4]
```diff
@@ -1,99 +1,191 @@
...
-from numpy.testing import assert_almost_equal
-
-#pytestmark = pytest.mark.skipif(True, reason='problem with the files')
-
-from tardis import __path__ as path
-
-test_path = os.path.join(path[0], 'montecarlo', 'test_montecarlo.so')
-
-tests = CDLL(test_path)
+from tardis.montecarlo.struct import *
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I think `import *` should be changed to import only the stuff that is used. That way you have more control and errors are easier to find.

Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Valid point.

**Comment by karandesai-96** [comment]
The structure of `compute_distance2continuum` is (with added debug statements):
```C
void compute_distance2continuum(rpacket_t * packet, storage_model_t * storage) {
  if (storage->cont_status == 0) {
      printf("A");
      //do job A
    } else {
      printf("B");
      // do job B
    }

  if (rpacket_get_virtual_packet(packet) > 0) {
      printf("C");
      // do job C
    } else {
      printf("D");
      // do job D
    }
}
```
For testing: 
* `model.d_cont` can be kept 0 / 1 to cover **A** and **B**.
* `packet.virtual_packet` can be kept 0 / 1 to cover **C** and **D**.

By adding all four combinations as testcases and executing the tests, `compute_distance2continuum` has been covered as:
`test_cmontecarlo.py ........**BD.AD.BC.AC.**.s.......XX.`

**Comment by karandesai-96** [comment]
Rebasing upon merge of #514 caused failing of tests. Tests have to be rectified.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 5]
```diff
@@ -138,11 +138,22 @@ def test_compute_distance2line(packet_params, expected_stat, expected_dist, pack
...
-    np.testing.assert_almost_equal(tests.test_compute_distance2continuum(),
-                                  distance_to_electron)
+@pytest.mark.parametrize(
+    ['packet_params', 'model_params', 'expected_params'],
+    [({'virtual_packet': 0}, {'cont_status': 0}, {'chi_cont': 6.652486e-16, 'd_cont': 4.359272608766106e+28}),
+     ({'virtual_packet': 0}, {'cont_status': 1}, {'chi_cont': 6.652486e-16, 'd_cont': 4.358643523052283e+28}),
+     ({'virtual_packet': 1}, {'cont_status': 0}, {'chi_cont': 6.652486e-16, 'd_cont': 1e+99}),
+     ({'virtual_packet': 1}, {'cont_status': 1}, {'chi_cont': 6.652486e-16, 'd_cont': 1e+99})]
+)
+def test_compute_distance2continuum(packet_params, model_params, expected_params, packet, model):
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Thanks for writing a test for that method. Currently continuum interaction is a feature in development. I'd say it is safe for now to only test `cont_status=0`. @chvogl Is there any reason to test `cont_status=1` in the current master?

Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Okay, I will remove the two test cases having `cont_status=1`

Comment (chvogl) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: @yeganer There is no reason to test cont_status=1 in the current master.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 6]
```diff
@@ -232,8 +232,29 @@ def test_increment_j_blue_estimator(packet_params, j_blue_idx, expected, packet,
...
 
 
-def test_montecarlo_line_scatter():
-    assert tests.test_montecarlo_line_scatter()
+@pytest.mark.parametrize(
+    ['packet_params', 'expected_params'],
+    # TODO: Add scientifically sound test cases.
+    [({'virtual_packet': 1, 'tau_event': 2.9e13, 'last_line': 0},
+      {'tau_event': 2.9e13, 'next_line_id': 2}),
+
```
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: @yeganer I am just dealing with numbers.

**Comment by karandesai-96** [comment]
Rebase after #533 got merged required changes in python struct - removal of `recently_crossed_boundary` flag.

File: tardis/montecarlo/struct.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+from ctypes import *
```
Comment (wkerzendorf) [inline tardis/montecarlo/struct.py]: This is bad! you should never do a `from xxx import *` - you will likely pollute your namespace with it.

Comment (karandesai-96) [inline tardis/montecarlo/struct.py]: Indeed, I am changing it right now.

**Comment by wkerzendorf** [comment]
@yeganer - how is this PR review coming along?

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 8]
```diff
@@ -1,99 +1,343 @@
...
+        'next_line_id': 1,
+        'last_line': 0,
+        'close_line': 0,
+        'current_continuum_id': 1,
+        'virtual_packet_flag': 1,
+        'virtual_packet': 0,
+        'status': 0,
+        'id': 0
+    }
+    return RPacket(**packet_default)
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I'd prefer writing  `RPacket({ ...})` and removing the variable `packet_defaul`. That would look cleaner to me.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 9]
```diff
@@ -1,99 +1,343 @@
...
+        'inner_boundary_albedo': 0.0,
+        'reflective_inner_boundary': 0,
+
+        'chi_bf_tmp_partial': (c_double * 20000)(*([160.0] * 20000)),
+        't_electrons': (c_double * 2)(*([0.0] * 2)),
+
+        'l_pop': (c_double * 20000)(*([2.0] * 20000)),
+        'l_pop_r': (c_double * 20000)(*([3.0] * 20000)),
+    }
+    return StorageModel(**model_default)
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Same as for `RPacket`.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 10]
```diff
@@ -1,99 +1,343 @@
...
+@pytest.fixture(scope="function")
+def mt_state():
+    """Fixture to return `RKState` object with default params initialized."""
+    mt_state_default = {
+        'key': (c_ulong * 624)(*([0] * 624)),
+        'pos': 0,
+        'has_gauss': 0,
+        'gauss': 0.0
+    }
+    return RKState(**mt_state_default)
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: See above.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 11]
```diff
@@ -1,99 +1,343 @@
...
 
-def test_compute_distance2line():
-	distance_to_line = 7.792353908000001e+17
-	tests.test_compute_distance2line.restype = c_double
-	assert_almost_equal(tests.test_compute_distance2line(),
-		distance_to_line)
+     ({'mu': -.3, 'r': 8.1e14}, {'inverse_time_explosion': 1 / 2.6e7},
+      1.0003117541351274)]
+)
+def test_rpacket_doppler_factor(packet_params, model_params, expected, packet, model):
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: In general I'd prefer `test_rpacket_doppler_factor(mu, r, inv_t_exp, expected):` That would make the tests more readable I think.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 12]
```diff
@@ -1,99 +1,343 @@
...
+        'current_continuum_id': 1,
+        'virtual_packet_flag': 1,
+        'virtual_packet': 0,
+        'status': 0,
+        'id': 0
+    }
+    return RPacket(**packet_default)
+
+
+@pytest.fixture(scope="function")
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I think there is an `autouse` keyword. We should use that for all 3 structs.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 13]
```diff
@@ -1,99 +1,343 @@
...
+     ({'nu_line': 0.2, 'next_line_id': 1, 'last_line': 0},
+      {'tardis_error': 0, 'd_line': 7.792353908000001e+17}),
+
+     ({'nu_line': 0.5, 'next_line_id': 1, 'last_line': 0},
+      {'tardis_error': 2, 'd_line': 0.0}),
+
+     ({'nu_line': 0.6, 'next_line_id': 0, 'last_line': 0},
+      {'tardis_error': 2, 'd_line': 0.0})]
+)
+def test_compute_distance2line(packet_params, expected_params, packet, model):
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I think we should test correctly returning errors in one test and the correct calculation of d_line in another. But this change can be done in a later iteration of these tests.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 8 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: - [x] Done.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 9 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: - [x] Done.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 10 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: - [x] Done.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 12 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: `TypeError: byref() argument must be a ctypes instance, not 'function'`  
Upon using `autouse`, this pops out and doesn't rectify, I tried some solutions but nothing works.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 11 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: - [x] Done.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 13 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I added a `TODO` at this stage.

**Comment by wkerzendorf** [comment]
@karandesai-96 nice work on adding some documentation to the beginning of the file.
@yeganer I'm signing off - if you're happy merge away.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 12 above)
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I have an idea on how it works but that's not a priority right now.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 14]
```diff
@@ -1,99 +1,340 @@
...
+    assert_almost_equal(packet.chi_cont, expected_params['chi_cont'])
+    assert_almost_equal(packet.d_cont, expected_params['d_cont'])
 
-def test_rpacket_doppler_factor():
-	doppler_factor = 0.9998556693818854
-	tests.test_rpacket_doppler_factor.restype = c_double
-	assert_almost_equal(tests.test_rpacket_doppler_factor(),
-		doppler_factor)
 
 @pytest.mark.skipif(True, reason='Bad test design')
```
Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I'd move this function blow the new tests. That way we have new & old nicely grouped together.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 12 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Okay, I will keep it in mind for future implementation.

File: tardis/montecarlo/tests/test_cmontecarlo.py (hunk 14 above)
Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Okay I will shift it in its original place.

Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: - [x] Done.

Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I was talking about `test_move_packet`. Sorry for not being clear.

Comment (karandesai-96) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Ah, I misunderstood it as "Move `test_rpacket_doppler_factor` below to group old and new tests nicely in **diffs**" :smile:
- [x] Done.

**Pull request merged**